Improve basicAutogradNotImplementedFallback + new tests

zou3519 opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #105592
* #105660
* __->__ #105591
* #105587

This PR:
- removes some reference count bumps (to potentially improve overhead)
- adds some tests for undefined gradients

zou3519 committed:
Improve basicAutogradNotImplementedFallback + new tests

This PR:
- removes some reference count bumps (to potentially improve overhead)
- adds some tests for undefined gradients

[ghstack-poisoned]

zou3519 committed:
Update on "Improve basicAutogradNotImplementedFallback + new tests"


This PR:
- removes some reference count bumps (to potentially improve overhead)
- adds some tests for undefined gradients

[ghstack-poisoned]

zou3519 committed:
Update on "Improve basicAutogradNotImplementedFallback + new tests"


This PR:
- removes some reference count bumps (to potentially improve overhead)
- adds some tests for undefined gradients

[ghstack-poisoned]

soulitzer commented on test/autograd/test_fallback.py:
```diff
@@ -315,6 +315,45 @@ def test_post_autograd_returns_leaf(self, mode):
+                return (
+                    torch._C._functions.UndefinedGrad()(a),
+                    torch._C._functions.UndefinedGrad()(b),
+                )
```
Are you trying to test the case where your outputs requires_grad=True?  In that case your hooks are registered on the outputs of this function, so they should fire before this happens?

zou3519 replied:
Good point

soulitzer reviewed (approved):
Cool optimization! 

For the test, maybe call UndefinedGrad after the op so the undefined grads flow through the hook? Or maybe I misunderstood what it is trying to test.

zou3519 committed:
Update on "Improve basicAutogradNotImplementedFallback + new tests"


This PR:
- removes some reference count bumps (to potentially improve overhead)
- adds some tests for undefined gradients

[ghstack-poisoned]

zou3519 commented:
@pytorchbot merge

soulitzer commented:
@pytorchbot merge
